File: 0421-Maximum XOR of Two Numbers in an Array/0421-Maximum XOR of Two Numbers in an Array.cpp
class Solution {
public:
    int findMaximumXOR(vector<int>& nums) {
        int result = 0;
        int mask = 0;
        for (int b = 30; b >= 0; --b)
        {
            mask |= (1 << b);
            unordered_set<int> table;
            for (int& num : nums)
            {
                table.insert(num & mask);
            }
            
            int candidate = result | (1 << b);
            for (int i : table)
            {
                if (table.find(candidate ^ i) != table.end())
                {
                    result = candidate;
                    break;
                }
            }
        }
        
        return result;
    }
};

// Trie
class TrieNode {
public:
    TrieNode() {
        next.resize(2);
    }
    vector<TrieNode*> next;
};

class Solution {
public:
    int findMaximumXOR(vector<int>& nums) {
        TrieNode* root = buildTrie(nums);
        int result = 0;
        for (int& num : nums)
        {
            result = max(result, findMaxXOR(root, num));
        }
        
        return result;
    }
    
    TrieNode* buildTrie(vector<int>& nums) {
        TrieNode* root = new TrieNode();
        TrieNode* node = root;
        for (int& num : nums)
        {
            node = root;
            for (int i = 30; i >= 0; --i)
            {
                int index = (num & (1 << i)) ? 1 : 0;
                if (node->next[index] == nullptr)
                {
                    node->next[index] = new TrieNode();
                }
                node = node->next[index];
            }
        }
        
        return root;
    }
    
    int findMaxXOR(TrieNode* root, int num)
    {
        TrieNode* node = root;
        int result = 0;
        for (int i = 30; i >= 0; --i)
        {
            int mask = 1 << i;
            int index = (num & mask) ? 0 : 1;
            if (node->next[index] != nullptr)
            {
                node = node->next[index];
                result |= mask;
            }
            else
            {
                node = node->next[1 - index];
            }
        }
        return result;
    }
};

File: 0401-0500/0421-Maximum XOR of Two Numbers in an Array/0421-Maximum XOR of Two Numbers in an Array.py
from functools import reduce
import collections

class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)

class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        root = TrieNode()
        for num in nums:
            node = root
            for i in range(30, -1, -1):
                index = (num >> i) & 1
                node = node.children[index]
        
        result = 0
        for num in nums:
            curr = 0
            node = root
            for i in range(30, -1, -1):
                index = (num >> i) & 1
                if 1 - index in node.children:
                    curr |= (1 << i)
                    node = node.children[1 - index]
                else:
                    node = node.children[index]
            result = max(result, curr)

        return result

class Solution:
    def findMaximumXOR(self, nums: 'List[int]') -> 'int':
        Trie = lambda: collections.defaultdict(Trie)
        trie = Trie()
        END = True
        for num in nums:
            reduce(dict.__getitem__, '{:031b}'.format(num), trie)[END] = True
            
        result = 0
        candidate = '01'
        for num in nums:
            curr = 0
            node = trie
            for i in range(30, -1, -1):
                mask = 1 << i
                j = 0 if num & mask else 1
                if candidate[j] in node:
                    curr |= mask
                    node = node[candidate[j]]
                else:
                    node = node[candidate[1 - j]]
            result = max(result, curr)
        return result
